lib/libutee/arch/arm/gprof/gprof.c

```diff
@@ -0,0 +1,342 @@
+/*
+ * Copyright (c) 2016, Linaro Limited
+ * All rights reserved.
+ *
+ * Redistribution and use in source and binary forms, with or without
+ * modification, are permitted provided that the following conditions are met:
+ *
+ * 1. Redistributions of source code must retain the above copyright notice,
+ * this list of conditions and the following disclaimer.
+ *
+ * 2. Redistributions in binary form must reproduce the above copyright notice,
+ * this list of conditions and the following disclaimer in the documentation
+ * and/or other materials provided with the distribution.
+ *
+ * THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS"
+ * AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
+ * IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
+ * ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT HOLDER OR CONTRIBUTORS BE
+ * LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR
+ * CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF
+ * SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS
+ * INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN
+ * CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
+ * ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
+ * POSSIBILITY OF SUCH DAMAGE.
+ */
+
+/*
+ * Portions of this file is adapted from glibc:
+ *   gmon/gmon.c
+ *   gmon/mcount.c
+ * Copyright (c)
+ *      The Regents of the University of California.  All rights reserved.
+ */
+
+#include <assert.h>
+#include <compiler.h>
+#include <inttypes.h>
+#include <malloc.h>
+#include <stdint.h>
+#include <string.h>
+#include <tee_api_private.h>
+#include <trace.h>
+#include <user_ta_header.h>
+#include <utee_syscalls.h>
+#include <utee_types.h>
+#include "gprof.h"
+
+static struct gmonparam _gmonparam = { GMON_PROF_OFF };
+
+/*
+ * Buffer for the pofile data. Should be allocated dynamically by malloc(),
+ * but since the TA heap size is defined by the TA itself via ta_heap and
+ * ta_heap_size, we'd better not interfere. Until we design another way,
+ * let's use a static buffer in .bss.
+ */
+static uint64_t _gprof_buf[256 * 1024 / sizeof(uint64_t)];
+
+static size_t _gprof_offs; /* Current offset in the output file */
+static uint32_t _gprof_file_id; /* File id returned by tee-supplicant */
+
+static void * _gprof_calloc(size_t size, size_t num)
+{
+	if ((num * size) > sizeof(_gprof_buf))
+		return NULL;
+	return _gprof_buf;
+}
+
+/* Adjust PC so that gprof can locate it in the TA ELF file */
+#define ADJUST_PC(pc) \
```

Comment:
This way of updating parameters to a macro is something I think should be avoided unless the alternative is to inconvenient, in this case it isn't.